Add support for user-space networking via pasta

lxp opened the pull request:
This pull request adds support for user-space networking via pasta ~~and enables it by default~~.

### Rational

Flatpak only supports sharing the network namespace with the host. This is especially problematic with abstract UNIX sockets (like the X11 abstract UNIX socket), since they can be accessed just by knowing the name.

For protecting against these attacks, we either need to fully block abstract UNIX sockets via seccomp or use a separate network namespace. We are going the route via a separate network namespace, since it is the better sandboxing approach. Futhermore, it provides more flexibility.

There are essentially two options to provide external network access in a separate namespace. Either use the kernel, which requires root access, since devices are added in the host network namespace. Or use user-space networking, which maps Layer-2 traffic from the sandbox to unprivileged Layer-4 (TCP, UDP, ...) sockets on the host.

One popular implementation of user-space networking is slirp4netns, which was initially used by Podman to provide rootless networking. However, it implements a full TCP/IP stack in user-space. Pasta is another user-space networking implementation, which tries to be more clever and avoids having to implement a full TCP/IP stack in user-space. Nowadays, Podman also uses pasta by default to provide rootless networking.

lxp commented:
I have re-based the pull request onto the current master branch.

Maybe I can add a bit more info regarding the option `config.pasta.mode`:

* `transparent`: Runs pasta mostly in default mode, which tries to emulate the host network as close as possible, i.e., it copies interface names, IP addresses and MAC addresses. It also allows access to services listening on localhost in the host network namespace and automatically exposes services listening in the sandbox to the outside world. Therefore, it mostly feels like being directly in the host network namespace.
* `isolate`: Restricts pasta as much as possible without any additional tools like iptables or nftables. Pasta determines the main network interface by looking at the default route in the host network namespace. This mode still allows full access to this main network. However, services running in the sandbox are not exposed to the host network namespace, since no incoming connections are forwarded. Also accessing services in the host network namespace, which are not listening on an IP address of the main network is not possible. This basically blocks access from the sandbox to services only listening on the loopback address in the host namespace.

max-privatevoid commented on modules/launch.nix:
```diff
@@ -50,7 +50,18 @@ let
   
+  pastaEnable = config.bubblewrap.network && config.pasta.enable;
+  originalBwrapExe = "${config.bubblewrap.package}/bin/bwrap";
+  bwrapExe = if pastaEnable then "${config.pasta.package}/bin/pasta" else originalBwrapExe;
```
This feels pretty hacky. It also seems to break the xdg-dbus-proxy wrapper currently. We also need to make sure that wrapping bwrap doesn't break the sync fd shenanigans.

lxp replied:
I agree with you that it is not the best solution. However, it was the one that I could get quickly working.

It might be possible to avoid having pasta as additional wrapper. However, this would require some major changes in the nixpak launcher.
Currently, it just sets up everything and hands over to bwrap by exec'ing it.
If we want to go down the other route, we would need bwrap to first setup the sandbox, then have the launcher adjust the network namespace with pasta before then bwrap actually executes the application within the sandbox.
This might be possible by using bwrap's `--block-fd` and pasta's *PID* option. However, we might have to somehow magically find the PID of the subprocess that bwrap forks (not sure if we can get that directly from bwrap or if we have to rely on the proc filesystem).

I think we might need to change the nixpak launcher anyway to something supervising the sandbox and, therefore, running while the whole sandbox lifetime.
I am working on a few more sandboxing features, namely for Wayland using `wayland-proxy-virtwl` and for PipeWire using `pw-container`.
Both tools have to run for the whole sandbox lifetime and have no mechanism like `xdg-dbus-proxy` to exit when an FD is closed.
I have basically both of these sandboxing features already working, but they currently leave dangling processes when the application exists.

What do you think about changing nixpak's launcher into a sandbox supervisor?
Would you prefer running pasta separately instead of as wrapper?

lxp replied:
If you want to take a sneak peak at my current state of Wayland and PipeWire sandboxing features: https://github.com/lxp/nixpak/pull/1
On my system, I am also running a re-implemented version of Snap's PipeWire access control scripts. However, I am not sure, if these changes can be integrated into nixpak, since they require changes to existing system services on the host system.

lxp replied:
I moved the pasta command construction into the Go launcher to fix the most obvious bug that the `--sync-fd` parameter is passed to pasta instead of bwrap.
However, it seems that pasta closes the sync fd used for xdg-dbus-proxy :(
I think the only way to proceed is to make nixpak's launcher into a sandbox supervisor and implement pasta support without wrapping the bwrap command like I described in my previous comment.
If you have different ideas, please let me know.

max-privatevoid commented on modules/pasta.nix:
```diff
@@ -0,0 +1,75 @@
+
+{
+  options.pasta = {
+    enable = mkEnableOption "Pasta networking" // { default = true; };
```
I don't think it should be enabled by default. It changes the network behavior quite a lot.

lxp replied:
I agree with you that it changes the behavior a lot. I have no problem with disabling it by default.
My main reason for suggesting to enable it, was to have a *secure-by-default* configuration.

max-privatevoid reviewed (changes requested):
Pretty cool. The fingerprinting resistance stuff is neat. But since this is a pretty invasive change to the "system" networking stack as perceived by the application, I think this should not be enabled by default.

Side note: We are now 3 layers deep in wrapper hell: `launcher -> pasta -> bwrap -> application`, and it's all hardcoded stuff. It might be time for me to come up with a more generic concept for defining (stackable) wrappers.

lxp commented:
> Pretty cool. The fingerprinting resistance stuff is neat. But since this is a pretty invasive change to the "system" networking stack as perceived by the application, I think this should not be enabled by default.

I dropped the commit that makes pasta default.

> Side note: We are now 3 layers deep in wrapper hell: `launcher -> pasta -> bwrap -> application`, and it's all hardcoded stuff. It might be time for me to come up with a more generic concept for defining (stackable) wrappers.

I have refactored the nixpak launcher into a sandbox supervisor and moved away from pasta being another wrapper.
Therefore, pasta can now also be used together with the dbus proxy.
@max-privatevoid It would be cool if you can take another look at this pull request.
If you want I can drop https://github.com/nixpak/nixpak/pull/93/commits/1843853e74b82795b8bb13a37144ae4b8a0fecf7 since it is not necessary anymore. However, it could also be useful in other places.

lxp commented:
I just pushed two more commits to refactor the config loading and command handling of bwrap, xdg-dbus-proxy and pasta into separate functions, since the launcher code somehow got complicated.

lxp commented:
I pushed a few more fixes for processes that fork into background or read stdin.
@max-privatevoid Do you have any feedback, if we can go forward with this change?
I am a bit hesitate to continue development without any feedback from your side.

I also did not rebase the merge request right now, since there are no relevant changes in the repository.
I can do that any time, just ping me.

max-privatevoid commented on launcher/main.go:
```diff
@@ -139,57 +172,343 @@ func readJsonArgs(filename string) (args []string) {
+	if err != nil {
+		panic(err)
+	}
+	if _, err := bwrapChild.Wait(); err != nil {
```
There seems to be a race condition here. I guess you'd need to check for `ECHILD`.
```console
$ nix build .#busybox
$ while ./result/bin/busybox echo ok; do :; done
ok
ok
ok
(...)
panic: waitid: no child processes

goroutine 1 [running]:
main.(*Bwrap).WaitUntilChildExit(0xc000170180?)
	github.com/nixpak/nixpak/launcher/main.go:386 +0x45
main.run()
	github.com/nixpak/nixpak/launcher/main.go:501 +0x3a5
main.main()
	github.com/nixpak/nixpak/launcher/main.go:507 +0x13
```

lxp replied:
Thanks for the re-producer!
Our new child reaper might catch the child exit before we explicitly wait for it. Since we do not need the exit code, I think it is fine to just let the child reaper catch it and simply ignore the error like you suggested, when we explicitly wait for the child and it already exited before.
I fixed it in the most recent version.

max-privatevoid reviewed (changes requested):
Sorry it took me so long to get back to you on this. I've been constrained in terms of how much time I can spend working on my projects, so I wasn't really motivated to review big PRs. I like what you've done so far, especially the launcher rework.

I've found 2 problems so far. One is the race condition on exit detailed in the other comment.

The other one I've only seen once and haven't been able to reproduce since. The first time I launched the pasta example, it complained that it couldn't join the pasta namespace, due to `EPERM`. Paraphrasing because I don't have the terminal output of that anymore. It never happened again after that, so I guess we can call it a fluke unless you happen to know about that specific behavior.

The stuff unrelated to the launcher looks good. And yeah, let's keep `sloth.uid` and `sloth.gid`.

I've taken the liberty to rebase and run `go fmt`.

lxp commented:
> Sorry it took me so long to get back to you on this. I've been constrained in terms of how much time I can spend working on my projects, so I wasn't really motivated to review big PRs. I like what you've done so far, especially the launcher rework.

I know the problem and thanks for taking the time to review my PR!

> I've found 2 problems so far. One is the race condition on exit detailed in the other comment.

I am going to look into it.

> I've taken the liberty to rebase and run `go fmt`.

Yeah, no problem. I also just did go through all commits and formatted them correctly. I hope there are no formatting problems with future commits, since I already started using the golang extension in VSCodium some time ago.

lxp commented:
> I've found 2 problems so far. One is the race condition on exit detailed in the other comment.

I fixed one of the concurrency issues you discovered.

I am still looking into the other one. It seems to be related to the following bubblewrap issue: https://github.com/containers/bubblewrap/issues/634
In some cases (e.g. when the `--dev` option is set) bubblewrap adds a second user namespace layer, which is only initialized after the info-fd is written (our indicator that we can start pasta). This leads to a race-condition and pasta either sees the first or second user namespace layer. Therefore, pasta currently only works when no bubblewrap `--dev` option is used or when it wins the race to see the first user namespace layer.
My current idea would be to find the correct namespace by querying the user namespace from the mount namespace via the kernel and enter this user namespace before launching pasta. Maybe pasta would also be open for improving their logic in finding the correct user namespace. However, I am still experimenting with this solution.

While experimenting with the second race condition, this also uncovered a faulty error handling: When pasta failed, the bubblewrap sandbox was not terminated correctly and it continued to run the application. I pushed a commit to fix this.

lxp commented:
> My current idea would be to find the correct namespace by querying the user namespace from the mount namespace via the kernel and enter this user namespace before launching pasta. Maybe pasta would also be open for improving their logic in finding the correct user namespace. However, I am still experimenting with this solution.

I just wrote a hacky patch for pasta that does exactly this and it works fine. I am currently talking to the developers of pasta if they want to integrate it.
Should we just ship the patch with NixPak for now until it is merged upstream?
As far as I know this is the last known issue and we could going in the direction of merging this MR.

max-privatevoid reviewed (approved):
I've been test driving this for a bit now and it seems to work well. Anything you wanna add? Otherwise I'll merge.

lxp commented:
By the way, this is the mailinglist thread where I am discussing how the pasta patch can be upstreamed: https://archives.passt.top/passt-user/671252c8-88f6-45b7-b719-b82786e84bb7@gnedt.at/
Current status: They are happy to accept it in some form. However, there is no decision yet if it will become the new default behavior or it will require additional options.

max-privatevoid commented:
Just bumped the version of the launcher since this is a pretty major change. Thank you for your incredible work.
